buildSrc/plugins/src/main/kotlin/org/jetbrains/kotlin/KotlinNativeTest.kt

```diff
@@ -0,0 +1,436 @@
+            // Apply closure set in build.gradle to get all parameters.
+            this.configure(config)
+            if (enabled) {
+                // Configure test task.
+                val target = project.testTarget
+                // If run task depends on something, compile task should also depend on this.
+                val compileTask = project.tasks.getByName("compileKonan${name.capitalize()}${target.name.capitalize()}")
+                compileTask.setSameDependenciesAs(this)
+                if (doBefore != null) compileTask.doFirst(doBefore!!)
+                compileTask.enabled = enabled
+            }
+        }
+
+/**
+ * Task to run tests compiled with TestRunner.
+ * Runs tests with GTEST output and parses it to create statistics info
+ */
+open class KonanGTest : KonanTest() {
+    init {
+        // Use GTEST logger to parse test results later
+        testLogger = Logger.GTEST
+        outputDirectory = "${project.testOutputStdlib}/$name"
+        val target = project.testTarget
+        executable = "$outputDirectory/${target.name}/$name.${target.family.exeSuffix}"
+    }
+
+    lateinit var statistics: Statistics
+
+    @TaskAction
+    override fun run() = runProcess(
+            executor = project.executor::execute,
+            executable = executable,
+            args = arguments
+    ).run {
+        statistics = parse(stdOut)
+        print()
+        check(exitCode == 0) { "Test $executable exited with $exitCode" }
+    }
+
+    private fun parse(output: String): Statistics = Statistics().apply {
+        Pattern.compile("\\[  PASSED  ] ([0-9]*) tests\\.").matcher(output)
+                .apply { if (find()) pass(group(1).toInt()) }
+
+        Pattern.compile("\\[  FAILED  ] ([0-9]*) tests.*").matcher(output)
+                .apply { if (find()) fail(group(1).toInt()) }
+        if (total == 0) {
+            // No test were run. Try to find if we've tried to run something
+            this.error(Pattern.compile("\\[={10}] Running ([0-9]*) tests from ([0-9]*) test cases\\..*")
+                    .matcher(output)
+                    .run { if (find()) group(1).toInt() else 1 })
+        }
+    }
+}
+
+/**
+ * Task to run tests built into a single predefined binary named `localTest`.
+ * Note: this task should depend on task that builds a test binary.
+ */
+open class KonanLocalTest : KonanTest() {
+    init {
+        // local tests built into a single binary with the known name
+        val target = project.testTarget
+        outputDirectory = project.testOutputLocal
+        executable = "$outputDirectory/${target.name}/localTest.${target.family.exeSuffix}"
+        testLogger = Logger.SILENT
+    }
+
+    @Optional
+    var expectedExitStatus = 0
+
+    /**
+     * Should this test fail or not.
+     */
+    @Optional
+    var expectedFail = false
+
+    /**
+     * Used to validate output as a gold value.
+     */
+    @Optional
+    lateinit var goldValue: String
```

Comment:
May be replace some of `lateinits` here and in other classes with just nullable properties? It looks like a null value is more suitable for expressing "there is no value for this property" than an uninitialized lateinit. So I suggest to keep only lateinits that _must_ be configured by a user (like `source`) and replace other ones with nullable properties.